README.md
# Homebridge Plugin for the Rinnai Touch WiFi Module
This Homebridge Plugin allows you to control a Brivis HVAC system via a Rinnai Touch WiFi Module. It supports the following operation modes:
* Single Temperature Set Point (ie. one controller with 1 to 5 zones including the Common zone)
* Multi Temperature Set Point (ie. one controller per zone, up to 4)

Functions available:
* Displaying the current state (eg. idle, heating, cooling)
* Switching to Off, Heating or Cooling modes
* Displaying the current temperature (depends on controller model)
* Setting the desired temperature
* Switching zones On and Off
* Switching the circulation fan On and Off as well as setting rotation speed
* Turning the water pump On and Off (for Evaporative Cooling only)
* Advancing to the next period of the Programme Schedule. (eg. Leave -> Return)
* Switching between Manual and Schedule control modes
* MQTT client. See [MQTT.md](docs/MQTT.md) for details.

## Accessories

This plugin will add one or more accessories to the Home app depending on your Rinnai Touch status. Accessories are discovered automatically without any need to modify the config.json file. The following table describes each type of accessory.

|Accessory|Description|
|-|-|
|Thermostat / Heater&nbsp;Cooler|Displays the current temperature, units (Celsius or Fahrenheit) and mode of the Brivis HVAC system. It allows you to set the desired temperature and change the mode. Modes are:<ul><li>`OFF` - System is off</li><li>`HEAT` - System is in heating mode</li><li>`COOL` - System is in cooling mode</li><li>`AUTO` - Returns system into Auto mode and the current period of the programme schedule (this option can be hidden with the `showAuto` config option). It will return to the `HEAT` or `COOL` mode when complete</li></ul>NOTES:<ul><li>One accessory will be added for each controller</li><li>Temperature units in the accessory do not determine which unit to use when displaying temperatures in the Home app. This is controlled by your phone's settings</li></ul>|
|Zone Switch|Shows if the zone is currently On or Off and allows you to change it. Zone Switches are shown if the operation mode is 'Single Temperature Set Point' and at least one zone is enabled (excluding the Common zone)<br/>NOTE: The 'Heater Cooler' accessory can be used as a zone switch|
|Fan|Displays the current state and speed setting of the circulation fan. Allows you to turn it Off or set the rotation speed<br/>NOTE: The fan can only be used when the Thermostat is in the `OFF` mode or `COOL` mode for Evaporative Cooling|
|Advance Period Switch|Shows if the Period of the Programme Schedule has been advanced and allows you to change it|
|Manual Switch|Shows if the Manual mode is On or Off and allows you to change it|
|Pump|Displays the current state of the pump if you have Evaporative Cooling. Allows you to turn it On or Off<br/>NOTE: The pump can only be used when the Thermostat is in `COOL` mode.|

## Installation
Note: This plugin requires homebridge to be installed first.

To install or upgrade to the latest version of this plugin:

    npm install -g homebridge-rinnai-touch-plugin@latest

## Configuration

This is a platform plugin that will register accessories and their services with the Bridge provided by homebridge. The plugin will attempt to discover your Rinnai Touch accessories automatically thus requiring zero configuration to the config.json file.

If you find the auto config is not correct for your system or some defaults are not to your liking there are some overrides you can define in the config.json file.

NOTE: Homebridge version 1.0.0 and onwards require an entry in the config.json file

|Option|Description|Default Value (if not supplied)|
|-|-|-|
|`platform`|Must be `"RinnaiTouchPlatform"`||
|`name`|The name of the platform|`"Rinnai Touch"`|
|`address`|IP Address of the WiFi module<br/>NOTE: leave blank for auto discovery||
|`port`|Port to use for the WiFI module<br/>NOTE: leave blank for auto discovery|`27847`|
|`controllerType`|The type of accessory to use for the controller(s). Options are:<br/>`T` for Thermostat<br/>`H` for Heater Cooler|`T`|
|`zoneType`|The type of accessory to use for controlling zones (only applicable for Single Temperature Set Point). Options are:<br>`N` for None (ie. don't show any accessory for zones<br/>`S` for Switch<br/>`H` for Heater Cooler|`S`|
|`showFan`|Show the fan accessory in the Home app|`true`|
|`showAuto`|Show the `AUTO` option in the Thermostat menu|`true`|
|`showAdvanceSwitches`|Show the Advance Period switch accessory in the Home app|`true`|
|`showManualSwitches`|Show the Manual switch accessory in the Home app|`true`|
|`closeConnectionDelay`|The time (ms) to wait for the TCP connection to fully close. Increasing this may reduce `Connection Refused` errors from occuring|`1100`|
|`connectionTimeout`|The time (ms) to wait to close the TCP connection after the last request. Set to `-1` to keep the connection open indefinitely, or `0` to close immediately|`5000`|
|`clearCache`|Clear all the plugin's cached accessories from homebridge to force full discovery of accessories on restart|`false`|
|`debug`|Output debug information to the Homebridge log|`false`|
|`mqtt`|See [MQTT.md](docs/MQTT.md) for details|`{}`|


#### Example: Bare mimimum (requried for Homebridge 1.0.0 onwards)

    "platforms": [
        {
            "platform": "RinnaiTouchPlatform"
        }
    ],


#### Example: Use the 'Heater Cooler' accessory and turn on debug logging

    "platforms": [
        {
            "platform": "RinnaiTouchPlatform",
            "name": "Rinnai Touch",
            "controllerType": "H",
            "debug": true
        }
    ],

#### Example: No Auto option, Advance Period & Manual switches
This is useful if you only use Manual Control of your HVAC (ie. no programme schedules).

    "platforms": [
        {
            "platform": "RinnaiTouchPlatform",
            "name": "Rinnai Touch",
            "showAuto": false,
            "showAdvanceSwitches": false,
            "showManualSwitches": false
        }
    ],

## Version History

See [Change Log](CHANGELOG.md).

## Known Limitations
* The plugin only supports a TCP connection over a LAN so no other connections can be active at the time. This would typically be the TouchApp by Rinnai.
* If the TCP connection is not closed properly then no further connections can be made to the module. I've tried to mitigate this as best I can by keeping TCP connections as short as possible and only allowing one request at a time. If it does happen I find rebooting my router clears it but rebooting the module itself should work also.
* Multi controller and Evaporative cooling configurations were not able to be tested so may not function properly.
* Due to the lag between sending a command to the module and it correctly reflecting that command in it's status there may be a short delay of a few seconds before the Home app shows the correct values. eg. When switching from HEAT to COOL mode some details such as the desired temperature will take a few seconds before the current value is shown.
* If the number of zones is different between the `Heat` and `Cool` modes the Zone Switches are dynamically added or removed as necessary. The downside of this is that you will loose any changes you made to the accessory (eg. name).
* If the WiFi module does not supply a current temperature then the temperature will display as zero in the Thermostat/Heater Cooler accessory. I would have prefered it showed as blank but couldn't find a way to do it. This appears to be a limitation of the service within Homebridge.
* The WiFi module will close the TCP connection after 5 minutes of inactivity. If the connection timeout is set to never (ie. `-1`) the plugin will attempt to automatically reconnect.
* The 'Heater Cooler' accessory is not currently supported by Home Assistant. See https://github.com/home-assistant/core/issues/30384


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## 2.5.2 (2020-07-03)

* [FIX] Initial MQTT publication may occur before connection to broker established

## 2.5.1 (2020-07-03)

* Simplify accessory settings
* Remove `AUTO` option from Zone 'Heater Cooler' accessory
* Additional INFO logging to show what caused the TCP connection to open

## 2.5.0 (2020-06-28)

* Allow 'Heater Cooler' accessory to be used as zone switches
* Add Current Temperature MQTT subscriptions for zones.
* [FIX] Fan does not switch off when set via MQTT hvac/mode/set topic
* [FIX] Handle case where status messages may be concatenated
* [FIX] Turning Zone Switch on while Thermostat & Fan are off throws error

## 2.4.4 (2020-06-21)

* [FIX] Setting mode via MQTT hvac/mode/set topic throws error

## 2.4.3 (2020-06-19)

* [FIX] Publish delay after processing command

## 2.4.2 (2020-06-18)

* Add 'retain' option to MQTT publish
* Improve time to detect command updates

## 2.4.1 (2020-06-17)

* New MQTT topics & status change publication event
* New config option for IP Address, Port & Connection Timeout
* Add support for Common zone
* Removal of Map Overrides

## 2.3.2 (2020-06-02)

* [FIX] Ignore decimal place when setting temperature via Home Assistant HVAC

## 2.3.1 (2020-06-02)

* Replace Simple MQTT format with Home Assistant HVAC specific one

## 2.3.0 (2020-05-26)

* MQTT support to publish status and subscribe to commands
* Add CHANGELOG file
* Further code cleanup
* Improved error handling and logging

## 2.2.0 (2020-05-17)

* Major code refactor
* New Manual Control switch
* Reintroduce Heater Cooler accessory as alternative to Thermostat accessory
* Show controller's temperature units in "Hardware Display Units" of Thermostat/Heater Cooler accessory
* Expose TCP close connection delay time as configuration option
* Add Config Schema file to allow configuration via Homebridge Config UI X

## 2.1.0 (2020-04-18)

* Set unique serial number for each accessory
* Set fan rotation direction
* Automatically switch fan off when switching HVAC on and vice versa
* Stability improvements when sending commands to WiFi module

## 2.0.0 (2020-04-11)

* New switch accessories for Fan, Pump & Advance Period
* Use Homebridge dynamic platform instead of single accessory
* Use Thermostat service instead of Heater Cooler
* Zero config option
* Better support for evaporative cooling
* Revamped mapping overrides

## 1.2.0 (2020-03-24)

* Better support for multiple controller setups

## 1.1.0 (2020-03-22)

* Automatic detection of HVAC options (add-on air con, evaporative cooling)
* Map overrides configuration
* Partial evaporative cooling support
* Retry TCP connection (useful when router is rebooted and IP address changes)
* Stability improvements

## 1.0.1 (2020-03-15)

* [FIX] Current State and Zones not working for mutli-controller setups

## 1.0.0 (2020-03-15)

* Add single accessory with Heater Cooler & Zone Switch services


index.js
const Settings = require('./util/Settings');
const Logger = require('./util/Logger');

const RinnaiTouchRepository = require('./repositories/RinnaiTouchRepository');
const RinnaiTouchService = require('./services/RinnaiTouchService');
const RinnaiTouchThermostat = require('./accessories/RinnaiTouchThermostat');
const RinnaiTouchHeaterCooler = require('./accessories/RinnaiTouchHeaterCooler');
const RinnaiTouchZoneSwitch = require('./accessories/RinnaiTouchZoneSwitch');
const RinnaiTouchFan = require('./accessories/RinnaiTouchFan');
const RinnaiTouchPump = require('./accessories/RinnaiTouchPump');
const RinnaiTouchAdvanceSwitch = require('./accessories/RinnaiTouchAdvanceSwitch');
const RinnaiTouchManualSwitch = require('./accessories/RinnaiTouchManualSwitch');

const NativeClient = require('./mqtt/NativeClient');
const HomeAssistantClient = require('./mqtt/HomeAssistantClient');
const TemperatureClient = require('./mqtt/TemperatureClient');

let Accessory, Service, Characteristic, UUIDGen;

/*  config.json
    {
        "platform": "RinnaiTouchPlatform",
        "name": "Rinnai Touch",
        "address": "192.168.1.3",
        "port": 27847,
        "controllerType": "T",
        "zoneType": "S",
        "showFan": true,
        "showAuto": true,
        "showAdvanceSwitches": true,
        "showManualSwitches": true;
        "closeConnectionDelay": 1100,
        "connectionTimeout": 5000,
        "clearCache": false,
        "debug": true,
        "mqtt": {}
    }
*/

module.exports = function(homebridge) {
    Accessory = homebridge.platformAccessory;
    Service = homebridge.hap.Service;
    Characteristic = homebridge.hap.Characteristic;
    UUIDGen = homebridge.hap.uuid;

    homebridge.registerPlatform("homebridge-rinnai-touch-plugin", "RinnaiTouchPlatform", RinnaiTouchPlatform, true);
}

class RinnaiTouchPlatform {
    #started = false;
    #repository;
    service;

    constructor(log, config, api) {
        try {
            this.settings = new Settings(config);
            this.log = new Logger(log, this.settings.debug);
    
            this.log.debug(this.constructor.name, undefined, 'log', JSON.stringify(config), 'api')
    
            this.Accessory = Accessory;
            this.Service = Service; 
            this.Characteristic = Characteristic;
            this.UUIDGen = UUIDGen;
    
            this.accessories = {};

            this.#repository = new RinnaiTouchRepository(this.log, this.settings);
            this.service = new RinnaiTouchService(this.log, this.#repository);

            this.#started = true;
    
            if (api) {
                this.api = api;
                this.api.on('didFinishLaunching', () => {
                    try {
                        this.discover();
                    }
                    catch(error) {
                        this.log.error(error);
                    }
                });

                this.api.on('shutdown', () => {
                    this.log.info('Shutting down plugin');
                    this.#repository.closeConnection();
                });
            }
        }
        catch(error) {
            log(`[ERROR] ${error.message}`);
        }
    }

    configureAccessory(accessory) {
        try {
            if (!this.#started)
                return;

            this.log.debug(this.constructor.name, 'configureAccessory', 'accessory');
            this.log.info(`Configure ${accessory.displayName}`);
    
            accessory.reachable = true;
    
            let rtAccessory;
    
            switch(accessory.context.type) {
                case 'thermostat':
                    rtAccessory = new RinnaiTouchThermostat(this);
                    break;
                case 'heatercooler':
                    rtAccessory = new RinnaiTouchHeaterCooler(this);
                    break;
                case 'zoneswitch':
                    rtAccessory = new RinnaiTouchZoneSwitch(this);
                    break;
                case 'fan':
                    rtAccessory = new RinnaiTouchFan(this);
                    break;
                case 'pump':
                    rtAccessory = new RinnaiTouchPump(this);
                    break;
                case 'advanceswitch':
                    rtAccessory = new RinnaiTouchAdvanceSwitch(this);
                    break;
                case 'manualswitch':
                    rtAccessory = new RinnaiTouchManualSwitch(this);
                    break;
            }
    
            rtAccessory.configure(accessory);
            this.accessories[rtAccessory.getKey()] = rtAccessory;    
        }
        catch(error) {
            this.log.error(error);
        }
    }

    async discover() {
        try {
            this.log.debug(this.constructor.name, 'discover');

            // Clear Cached accessories if required
            if (this.settings.clearCache) {
                let accessories = Object.values(this.accessories).map((acc) => acc.accessory);
                this.api.unregisterPlatformAccessories("homebridge-rinnai-touch-plugin", "RinnaiTouchPlatform", Object.values(accessories));
                this.accessories = {};
            }

            await this.service.init();
   
            if (this.service.hasHeater) this.log.info('Found Heater');
            if (this.service.hasCooler) this.log.info('Found Cooler');
            if (this.service.hasEvaporative) this.log.info('Found Evaporative Cooler');
            
            let zones = this.service.zones.map(z => this.service.getZoneName(z));
            this.log.info(`Found Zone(s): ${zones.join()}`);

            let operation = this.service.hasMultiSetPoint ? 'Multi' : 'Single';
            this.log.info(`Operation Mode: ${operation} Temperature Set Point`);

            this.configureThermostats();
            this.configureHeaterCoolers();
            this.configureZoneSwitches();
            this.configureFan();
            this.configureAdvanceSwitches();
            this.configureManualSwitches();
            this.configurePump();

            this.service.on('mode', (mode) => {
                if (!this.service.hasMultiSetPoint) {
                    this.configureZoneSwitches();
                    this.configureHeaterCoolers();
                }
                if (this.service.hasEvaporative) {
                    this.configurePump();
                }
            });

            if (this.settings.mqtt.formatNative) {
                this.nativeClient = new NativeClient(this.log, this.settings.mqtt, this.#repository);
            }
            if (this.settings.mqtt.formatHomeAssistant) {
                this.homeAssistantClient = new HomeAssistantClient(this.log, this.settings.mqtt, this.service);
            }
            if (this.settings.mqtt.subscribeTemperature) {
                this.temperatureClient = new TemperatureClient(this.log, this.settings.mqtt, this.service);
            }
        }
        catch(error) {
            this.log.error(error);
        }
    }

    configureThermostats() {
        this.log.debug(this.constructor.name, 'configureThermostats');

        for(let zone of this.service.AllZones) {
            if (this.settings.controllerType === 'T' && this.service.controllers.includes(zone)) {
                let name = this.service.hasMultiSetPoint
                    ? this.service.getZoneName(zone)
                    : this.settings.name;
                this.addAccessory(RinnaiTouchThermostat, name, zone);
            } else {
                this.removeAccessory(RinnaiTouchThermostat, zone);
            }
        }
    }

    configureHeaterCoolers() {
        this.log.debug(this.constructor.name, 'configureHeaterCoolers');

        for(let zone of this.service.AllZones) {
            let addHeaterCooler =
                (this.settings.controllerType === 'H' && this.service.controllers.includes(zone)) ||
                (!this.service.hasMultiSetPoint && zone !== 'U' && this.settings.zoneType === 'H' && this.service.zones.includes(zone));

            if (addHeaterCooler) {
                let name = zone !== 'U'
                    ? this.service.getZoneName(zone)
                    : this.settings.name;
                this.addAccessory(RinnaiTouchHeaterCooler, name, zone);
            } else {
                this.removeAccessory(RinnaiTouchHeaterCooler, zone);
            }
        }
    }

    configureZoneSwitches() {
        this.log.debug(this.constructor.name, 'configureZoneSwitches');

        for(let zone of ['A','B','C','D']) {
            if (!this.service.hasMultiSetPoint && this.settings.zoneType === 'S' && this.service.zones.includes(zone)) {
                this.addAccessory(RinnaiTouchZoneSwitch, this.service.getZoneName(zone), zone);
            } else {
                this.removeAccessory(RinnaiTouchZoneSwitch, zone);
            }
        }
    }

    configureFan() {
        this.log.debug(this.constructor.name, 'configureFan');

        if (this.settings.showFan) {
            this.addAccessory(RinnaiTouchFan, 'Circulation Fan');
        } else {
            this.removeAccessory(RinnaiTouchFan);
        }
    }

    configureAdvanceSwitches() {
        this.log.debug(this.constructor.name, 'configureAdvanceSwitch');

        for(let zone of this.service.AllZones) {
            if (this.settings.showAdvanceSwitches && this.service.controllers.includes(zone)) {
                let name = this.service.hasMultiSetPoint ? `Advance Period ${zone}` : 'Advance Period';
                this.addAccessory(RinnaiTouchAdvanceSwitch, name, zone);
            } else {
                this.removeAccessory(RinnaiTouchAdvanceSwitch, zone);
            }
        }
    }

    configureManualSwitches() {
        this.log.debug(this.constructor.name, 'configureManualSwitch');

        for(let zone of this.service.AllZones) {
            if (this.settings.showManualSwitches && this.service.controllers.includes(zone)) {
                let name = this.service.hasMultiSetPoint ? `Manual ${zone}` : 'Manual';
                this.addAccessory(RinnaiTouchManualSwitch, name, zone);
            } else {
                this.removeAccessory(RinnaiTouchManualSwitch, zone);
            }
        }
    }

    configurePump() {
        this.log.debug(this.constructor.name, 'configurePump');

        if (this.service.hasEvaporative) {
            this.addAccessory(RinnaiTouchPump, 'Evaporative Pump');
        } else {
            this.removeAccessory(RinnaiTouchPump);
        }
    }

    addAccessory(RinnaiTouchAccessory, name, zone) {
        this.log.debug(this.constructor.name, 'addAccessory', RinnaiTouchAccessory.name, name, zone);
      
        let rtAccessory = new RinnaiTouchAccessory(this);
        let key = rtAccessory.getKey(zone);
        if (key in this.accessories) {
            return;
        }

        rtAccessory.init(name, zone);

        this.accessories[key] = rtAccessory;
        this.api.registerPlatformAccessories("homebridge-rinnai-touch-plugin", "RinnaiTouchPlatform", [rtAccessory.accessory]);
        this.log.info(`Add ${rtAccessory.accessory.displayName}`);
    }

    removeAccessory(RinnaiTouchAccessory, zone) {
        this.log.debug(this.constructor.name, 'removeAccessory', RinnaiTouchAccessory.name, zone);

        let rtAccessory = new RinnaiTouchAccessory(this);
        let key = rtAccessory.getKey(zone);
        if (!(key in this.accessories))
            return;
    
        let accessory = this.accessories[key].accessory;
        this.api.unregisterPlatformAccessories("homebridge-rinnai-touch-plugin", "RinnaiTouchPlatform", [accessory]);
        delete this.accessories[key];
        this.log.info(`Remove ${accessory.displayName}`);
    }
}